crates/bevy_reflect/src/lib.rs

```diff
@@ -26,15 +26,60 @@
 //! there are some [limitations] with this crate.
 //!
-//! # The `Reflect` Trait
+//! # The `Reflect` and `PartialReflect` traits
 //!
-//! At the core of [`bevy_reflect`] is the [`Reflect`] trait.
+//! At the root of [`bevy_reflect`] is the [`PartialReflect`] trait.
 //!
-//! One of its primary purposes is to allow all implementors to be passed around
-//! as a `dyn Reflect` trait object.
-//! This allows any such type to be operated upon completely dynamically (at a small [runtime cost]).
+//! Its purpose is to allow dynamic [introspection] of values,
+//! following Rust's type system through a system of [subtraits].
 //!
-//! Implementing the trait is easily done using the provided [derive macro]:
+//! Its primary purpose is to allow all implementors to be passed around
+//! as a `dyn PartialReflect` trait object in one of the following forms:
+//! * `&dyn PartialReflect`
+//! * `&mut dyn PartialReflect`
+//! * `Box<dyn PartialReflect>`
+//!
+//! This allows values of types implementing `PartialReflect`
+//! to be operated upon completely dynamically (at a small [runtime cost]).
+//!
+//! Building on `PartialReflect` is the [`Reflect`] trait.
+//!
+//! `PartialReflect` is a supertrait of `Reflect`
+//! so any type implementing `Reflect` implements `PartialReflect` by definition.
+//! `dyn Reflect` trait objects can be used similarly to `dyn PartialReflect`,
+//! but `Reflect` is also often used in trait bounds (like `T: Reflect`).
+//!
+//! The distinction between `PartialReflect` and `Reflect` is summarised in the following:
+//! * `PartialReflect` is a trait for interacting with values under `bevy_reflect`'s data model.
+//!   This means values implementing `PartialReflect` can be dynamically constructed and introspected.
+//! * The `Reflect` trait, however, ensures that the interface exposed by `PartialReflect`
+//!   on types which additionally implement `Reflect` mirrors the structure of a single Rust type.
+//! * This means `dyn Reflect` trait objects can be directly downcasted to concrete types,
+//!   where `dyn PartialReflect` trait object cannot.
+//! * `Reflect`, since it provides a stronger type-correctness guarantee,
+//!   is the trait used to interact with [the type registry].
+//!
+//! ## Converting between `PartialReflect` and `Reflect`
+//!
+//! Since `T: Reflect` implies `T: PartialReflect`, conversion from a `dyn Reflect` to a `dyn PartialReflect`
+//! trait object (upcasting) is infallible and can be performed with one of the following methods.
+//! Note that these are temporary while [the language feature for dyn upcasting coercion] is experimental:
+//! * [`PartialReflect::as_partial_reflect`] for `&dyn PartialReflect`
+//! * [`PartialReflect::as_partial_reflect_mut`] for `&mut dyn PartialReflect`
+//! * [`PartialReflect::into_partial_reflect`] for `Box<dyn PartialReflect>`
+//!
+//! For conversion in the other direction --- downcasting `dyn PartialReflect` to `dyn Reflect`,
```

Comment:
Terribly Minor Nit: Does the `---` actually render to a dash? If not, you could use the character directly:
```suggestion
//! For conversion in the other direction — downcasting `dyn PartialReflect` to `dyn Reflect`,
```

Also, I'm not a strong grammarist haha, but shouldn't you close out with another dash before continuing the sentence?
```suggestion
//! For conversion in the other direction — downcasting `dyn PartialReflect` to `dyn Reflect` —
```